# Calculation of RSE and quantitative determination of PFAS concentrations from the LC-TQ
Author: Valerie de Rijk

Date: 15/8/2025

## 1 Data Loading
First we load in the raw data as a csv file (which is the file you directly get from MassHunter).

In [11]:
%run src/functions_quantitative.ipynb
rawdata = load_and_label_pfas_csv("rawdata/rawdatatrial_allheaders.csv")

['PFBA', '13C4-PFBA', 'PFPeA', '13C5-PFPeA', 'HFPO-DA', '13C3-HFPO-DA', 'PFBS', '13C3-PFBS', 'PFHxA', '13C5-PFHxA', '4:2 FTS', '13C2-4:2FTS', 'PFPeS', 'PFPHpA', '13C4-PFHpA', 'ADONA', 'PFHxS', '13C3-PFHxS', 'PFOA', '13C8-PFOA', '6:2 FTS', '13C2-6:2FTS', 'PFHpS', 'PFNA', '13C9-PFNA', 'PFOSA', 'PFOS', '13C8-PFOSA', '13C8-PFOS', 'NMeFOSA', 'PFDA', 'D3-NMeFOSA', '13C6-PFDA', 'NEtFOSA', '8:2 FTS', '13C2-8:2FTS', '9Cl-PF3ONS', 'D5-NEtFOSA', 'PFNS', 'PFUnA', 'NMeFOSAA', '13C7-PFUnA', 'D3-NMeFOSAA', 'NEtFOSAA', 'D5-NEtFOSAA', 'PFDS', 'PFDoA', '13C2-PFDoA', 'NMeFOSE', 'D7-NMeFOSE', 'NEtFOSE', '11Cl-PF3OUdS', 'D9-NEtFOSE', 'PFTrDA', 'PFTeDA', '13C2-PFTeDA']
Total compounds: 56
Without isotope prefix: 32


## 2 Computing RSE for EIS compounds and target analytes (Instrument linearity)
We compute the RSE based on  the formula for RSE based on EPA method 1633 (section 10.3.3.3, page 28, option 2). The functions are loaded in a different script. 
If the RSE remains above 20%, this means something is wrong with the linearity of the LC-MS TQ and the calculated concentrations in section 3 are not trustworthy. 

*We mute the output of this cell for clarity of the script and summarize them under EIS results and Target Analytes results. If interested in full computation, please unmute the cell by removing the cell line **%%capture***

The output is structured as following: 
- List of identified EIS compounds (check if all are found)
- Computation for the EIS compounds
- Computation for the target analytes
- Summary of the result for EIS
- Summary of the result for target

*note: if not interested in detailed calculations, run the cell and ignore the output*


In [29]:
%%capture
%run src/functions_quantitative.ipynb
import importlib
import expected_concs
from expected_concs import expected_concs_EIS, expected_concs_target_analytes
# Example run
rse_results_eis = calculate_eis_rse(rawdata, expected_concs_EIS)
rse_results_target = calculate_target_rse(rawdata, expected_concs_target_analytes)

print_rse_summary(rse_results_eis)
print_rse_summary(rse_results_target)

### 2.1 EIS results

We print the results in a clear table and save them to csv (first for EIS, then target). Give the files an intuitive name.

In [13]:
df_rse_eis=df_rse_results(rse_results_eis)
display(df_rse_eis)

Compound  RSE_Percent  ...  Pass_Fail Sum_Relative_Squared_Error
0      13C4-PFBA     0.725334  ...       PASS                   0.000053
1     13C5-PFPeA     1.597641  ...       PASS                   0.000255
2   13C3-HFPO-DA     4.346961  ...       PASS                   0.001890
3      13C3-PFBS     4.889447  ...       PASS                   0.002391
4     13C5-PFHxA     3.128308  ...       PASS                   0.000979
5    13C2-4:2FTS    74.690995  ...       FAIL                   0.557874
6     13C4-PFHpA     2.509613  ...       PASS                   0.000630
7     13C3-PFHxS     6.164158  ...       PASS                   0.003800
8      13C8-PFOA     7.587856  ...       PASS                   0.005758
9    13C2-6:2FTS    79.200956  ...       FAIL                   0.627279
10     13C9-PFNA     6.171349  ...       PASS                   0.003809
11    13C8-PFOSA     4.974712  ...       PASS                   0.002475
12     13C8-PFOS     5.462934  ...       PASS                   0.002984
13    D3-NMeFOSA     3.171957  ...       PASS                   0.001006
14     13C6-PFDA     6.388374  ...       PASS                   0.004081
15   13C2-8:2FTS    71.672529  ...       FAIL                   0.513695
16    D5-NEtFOSA     5.557040  ...       PASS                   0.003088
17    13C7-PFUnA    10.639992  ...       PASS                   0.011321
18   D3-NMeFOSAA     4.570409  ...       PASS                   0.002089
19   D5-NEtFOSAA    11.124609  ...       PASS                   0.012376
20    13C2-PFDoA     9.649420  ...       PASS                   0.009311
21    D7-NMeFOSE     6.500790  ...       PASS                   0.004226
22    D9-NEtFOSE     7.976893  ...       PASS                   0.006363
23   13C2-PFTeDA     9.518922  ...       PASS                   0.009061

[24 rows x 7 columns]

### 2.2 Target Analytes results

In [14]:
df_rse_target = df_rse_results(rse_results_target)
display(df_rse_target)

Compound  RSE_Percent  ...  Pass_Fail Sum_Relative_Squared_Error
0           PFBA    14.699247  ...       PASS                   0.021607
1          PFPeA     7.422499  ...       PASS                   0.005509
2        HFPO-DA    17.089360  ...       PASS                   0.029205
3           PFBS    12.590153  ...       PASS                   0.015851
4          PFHxA    13.739342  ...       PASS                   0.018877
5        4:2 FTS    31.855531  ...       FAIL                   0.101477
6          PFPeS    11.295795  ...       PASS                   0.012759
7          ADONA    14.244006  ...       PASS                   0.020289
8          PFHxS    50.559110  ...       FAIL                   0.255622
9           PFOA    16.319428  ...       PASS                   0.026632
10       6:2 FTS    30.018149  ...       FAIL                   0.090109
11         PFHpS    35.305090  ...       FAIL                   0.124645
12          PFNA    48.002911  ...       FAIL                   0.230428
13         PFOSA    15.071720  ...       PASS                   0.022716
14          PFOS    10.250958  ...       PASS                   0.010508
15       NMeFOSA    14.802678  ...       PASS                   0.021912
16          PFDA     9.191516  ...       PASS                   0.008448
17       NEtFOSA    51.757747  ...       FAIL                   0.267886
18       8:2 FTS    35.641343  ...       FAIL                   0.127031
19    9Cl-PF3ONS    12.514639  ...       PASS                   0.015662
20          PFNS    13.557720  ...       PASS                   0.018381
21         PFUnA    23.468254  ...       FAIL                   0.055076
22      NMeFOSAA    24.958094  ...       FAIL                   0.062291
23      NEtFOSAA    44.708817  ...       FAIL                   0.199888
24          PFDS    14.650004  ...       PASS                   0.021462
25         PFDoA    19.505615  ...       PASS                   0.038047
26       NMeFOSE   233.253560  ...       FAIL                   5.440722
27       NEtFOSE   241.192406  ...       FAIL                   5.817378
28  11Cl-PF3OUdS   103.188367  ...       FAIL                   1.064784
29        PFTrDA    12.418280  ...       PASS                   0.015421
30        PFTeDA    11.758573  ...       PASS                   0.013826

[31 rows x 7 columns]

## 3 Instrument checks
*all of these calculations still need to be added*

We perform different checks apart from instrument linearity to ensure that the LC-TQ is behaving as we expect. 
1. Instrument Sensitivity check (signal to noise ratio)
     - here we check if the ratio is greater than or equal to 3:1 (EPA section 14.1)
2. Bile salt interference check
   - We check that the RT of bile salts falls at least one minute outside of the retention time for any linear or branched PFOS (EPA 14.2)
3. Calibration verification (EPA 14.3)
   - Check that the CV, e.g. the mid-level calibration point, (which should be done every 10 samples or less) has a recovery between 70 and 130%.
4. Retention times
    - All target analyte with  true isotope dilution should elute within 0.1 minutes of the associated EIS compounds.
5. Ongoing precision and recovery
    - The recovery of native compounds and EIS compounds must meet the standards in table 5,6,7 and 8 of the EPA.
6. Instrument blank and method blank
   - Check that is it below all analyte's MDL.
7. Instrument sensitivity
   - This is the ratio of the NIS peak areas from QC and field samples relative to the mean area of the corresponding NIS in most recent inital calibration. This must be within 50 to 200%. 


## 3.1 recovery  EIS compounds
This follows EPA method 1633 page 44 and EPA method page 28 (equation for RFs). We first compute the average response factor for the EIS compounds in the calibration standard. Then, we use the formula on page 44 to calculate the concentration in the sample and compare it to the expected concentration to determine the recovery. 

### 3.1.1 Calculating average Response Factor EIS compounds

In [15]:
from analogs_compounds import target_EIS_analogs, EIS_NIS_analogs
from solutions import calibration_solutions

eis_RFS = calculate_average_RFs_EIS(rawdata, EIS_NIS_analogs, calibration_solutions)


AVERAGE RESPONSE FACTORS FOR EIS COMPOUNDS (n=24)


Compound_Name  NIS_Analog  ...  Average_Response_Factor  RSD_%
0     13C7-PFUnA   13C2-PFDA  ...                   1.6198   8.99
1     D3-NMeFOSA   13C4-PFOS  ...                   1.1955   2.68
2    13C2-8:2FTS  18O2-PFHxS  ...                   0.1306  54.19
3    13C2-6:2FTS  18O2-PFHxS  ...                   0.2116  59.87
4    D3-NMeFOSAA   13C4-PFOS  ...                   0.2326   3.86
5     13C5-PFPeA  13C2-PFHxA  ...                   0.6797   1.35
6     13C8-PFOSA   13C4-PFOS  ...                   4.1090   4.20
7    D5-NEtFOSAA   13C4-PFOS  ...                   0.2048   9.40
8     13C5-PFHxA  13C2-PFHxA  ...                   1.0973   2.64
9      13C6-PFDA   13C2-PFDA  ...                   1.8814   5.40
10    D7-NMeFOSE   13C4-PFOS  ...                   0.5773   5.49
11     13C9-PFNA   13C5-PFNA  ...                   0.9888   5.22
12   13C2-PFTeDA   13C2-PFDA  ...                   3.4117   8.04
13     13C3-PFBS  18O2-PFHxS  ...                   2.1140   4.13
14    13C4-PFHpA  13C2-PFHxA  ...                   0.8986   2.12
15    D5-NEtFOSA   13C4-PFOS  ...                   1.2825   4.70
16    D9-NEtFOSE   13C4-PFOS  ...                   0.6388   6.74
17     13C8-PFOS   13C4-PFOS  ...                   1.0407   4.62
18  13C3-HFPO-DA  13C2-PFHxA  ...                   0.4454   3.67
19     13C4-PFBA   13C3-PFBA  ...                   1.1364   0.61
20     13C8-PFOA   13C4-PFOA  ...                  13.0621   6.41
21   13C2-4:2FTS  18O2-PFHxS  ...                   0.4950  56.46
22    13C2-PFDoA   13C2-PFDA  ...                   2.0610   8.16
23    13C3-PFHxS  18O2-PFHxS  ...                   1.2772   5.21

[24 rows x 6 columns]

### 3.1.2 Concentration and recovery of EIS compounds 
*TO DO the weights of the bottles should be added as a column per sample*
*TO DO make a check if recovery falls in Table 5 + add RSD*


Here we compute the concentration of the EIS compounds based on the calculated response factor, the dilution factor and the sample volume (EPA 1633, page 44). 
It is important to enter the correct sample volume, otherwise calculations will be wrong. 

In [30]:
# computation of EIS concentration and recovery rate
from solutions import NIS_stock, EIS_stock
from allowed_recoveries import EIS_NIS_compound_recovery_aqueous
DF_sample = 1 #dilution factor 
mass_bottle_filled = 499.8 #grams
mass_bottle_empty = 53.68 #grams
Ws = (mass_bottle_filled - mass_bottle_empty)/1000 #L

dilution_NIS_stock = 2 #1:1 dilution
spiked_amount = 0.05 #ml

dilution_EIS_stock = 2 
spiked_amount_EIS =0.05 #ml

calculated_cons_EIS = calculate_conc_EIS(rawdata, eis_RFS, dilution_NIS_stock, Ws,EIS_NIS_analogs, DF_sample, NIS_stock, spiked_amount)

calculated_recoveries_EIS = calculate_recoveries_EIS(calculated_cons_EIS, EIS_stock, dilution_EIS_stock, spiked_amount_EIS, Ws)
# Run the validation
validation_results = validate_recoveries(calculated_recoveries_EIS, EIS_NIS_compound_recovery_aqueous)

# Get summary
summary = summarize_validation_results(validation_results)
print("Validation Summary:")
for key, value in summary['Overall'].items():
    print(f"{key}: {value}")

failed_recoveries = get_failed_recoveries(validation_results)

Name    Type  ...  13C2-PFDoA_Conc_Calc  13C3-PFHxS_Conc_Calc
0           blank + EIS  Sample  ...             15.583309             28.438385
1                QC std  Sample  ...             12.800405             29.063622
2                   KCS  Sample  ...             13.743221             27.727590
3                   DWZ  Sample  ...             20.234170             42.129966
4                   KWZ  Sample  ...             15.908230             35.223846
5                   DWP  Sample  ...             11.411672             25.897951
6                   KWP  Sample  ...             14.249636             34.258568
7                   KCS  Sample  ...             12.896930             29.357592
8                 QCstd  Sample  ...             14.317611             28.387927
9                 rinse  Sample  ...              0.597502             48.722371
10                rinse  Sample  ...              4.638626             21.319465
11                rinse  Sample  ...             10.712208            232.938959
12  blank_spike_eis-nis  Sample  ...             16.228393             30.722143
13                1_ccf  Sample  ...             11.956153             25.790454
14                 2_cf  Sample  ...             11.514315             25.037976
15                  3_f  Sample  ...             12.088705             25.799229
16               4_none  Sample  ...             12.648820             27.387549
17                rinse  Sample  ...              4.186591             53.230655
25                RINSE  Sample  ...              1.788427             97.672898
26          blank_spike  Sample  ...             15.719259             28.971481
27             alex_kwz  Sample  ...             12.553655             26.088812
28             alex_dwz  Sample  ...              9.073326             19.882518
29          alex_kwpfas  Sample  ...              8.849688             21.047432
30          alex_dwpfas  Sample  ...             10.304357             22.352751

[24 rows x 26 columns]

Name    Type  ...  13C2-PFDoA_Recovery_%  13C2-PFTeDA_Recovery_%
0           blank + EIS  Sample  ...             111.232411              104.757159
1                QC std  Sample  ...              91.368268               88.337193
2                   KCS  Sample  ...              98.098011              100.940880
3                   DWZ  Sample  ...             144.429888              107.225137
4                   KWZ  Sample  ...             113.551674               88.715303
5                   DWP  Sample  ...              81.455604               67.256793
6                   KWP  Sample  ...             101.712761               74.449688
7                   KCS  Sample  ...              92.057254               90.541372
8                 QCstd  Sample  ...             102.197962               98.666354
9                 rinse  Sample  ...               4.264920                6.328223
10                rinse  Sample  ...              33.110141               11.275904
11                rinse  Sample  ...              76.462882               81.633790
12  blank_spike_eis-nis  Sample  ...             115.836972              109.191898
13                1_ccf  Sample  ...              85.342067               83.750037
14                 2_cf  Sample  ...              82.188258               74.223994
15                  3_f  Sample  ...              86.288210               78.409591
16               4_none  Sample  ...              90.286269               92.698655
17                rinse  Sample  ...              29.883552               38.772459
25                RINSE  Sample  ...              12.765648                4.891989
26          blank_spike  Sample  ...             112.202814              116.293930
27             alex_kwz  Sample  ...              89.606987               88.961515
28             alex_dwz  Sample  ...              64.764677               58.598353
29          alex_kwpfas  Sample  ...              63.168365               56.218308
30          alex_dwpfas  Sample  ...              73.551674               67.315415

[24 rows x 26 columns]

Validation Summary:
Total_Measurements: 576
Passed: 472
Failed: 80
No_Data: 0
No_Reference: 24
Pass_Rate_%: 85.5072463768116


Sample_Name Sample_Type     Compound  ...  Expected_Upper_%  Status  Compound_Type
74          DWZ      Sample    13C4-PFBA  ...             130.0    Fail            EIS
76          DWZ      Sample   13C5-PFPeA  ...             130.0    Fail            EIS
82          DWZ      Sample   13C5-PFHxA  ...             130.0    Fail            EIS
83          DWZ      Sample   13C4-PFHpA  ...             130.0    Fail            EIS
84          DWZ      Sample    13C8-PFOA  ...             130.0    Fail            EIS
..          ...         ...          ...  ...               ...     ...            ...
449       RINSE      Sample   D3-NMeFOSA  ...             130.0    Fail            EIS
451       RINSE      Sample    13C9-PFNA  ...             130.0    Fail            EIS
452       RINSE      Sample    13C6-PFDA  ...             130.0    Fail            EIS
453       RINSE      Sample   13C7-PFUnA  ...             130.0    Fail            EIS
455       RINSE      Sample  13C2-PFTeDA  ...             130.0    Fail            EIS

[80 rows x 8 columns]

## 4 Quantitative determination of concentrations target analytes
In this section we will compute the concentration of all the analyzed PFAS compounds in the sample. Our current method is capable of analyzing 40 different compounds. Here, we distinguish between two categories to perform the calculation: 
- True Isotope Dilution Quantification (ID): 24 compounds are directly compared to their istopically labeled analog.
- Extracted Internal Standard Quantification (EIS): 16 compounds are quantified by comparing to a *different* isotopically labeled compound.

### 4.1 Response Ratio (RR) and Response Factor (RF) for target ID compounds and target EIS compounds.
Here we compute the response ratio for the ID compounds by using the formula from EPA 1633 on page 27 under section 10.3.3.2. For now we also fix some naming errors in the raw data for the target analytes. The only difference with the above calculation is that EIS analogs instead of direct pairs are used for the target EIS compounds. For ease, we call both RF.

- *Note 1*: NMeFoSA and NEtFoSa are calculated now but are not included in the PFAC30PAR ampoule we currently use in the lab, so these compounds cannot be in the calibration line. Use these results with care.
- *Note 2*:EPA 1633 mentions that PFTrDA recovery should improve by taking the average of the analog EIS compounds 13c2-PFTeDA and 13c2-PFDoa. Currently it only takes the EIS compound 13C2-PFTeDA as an analog.
- *Note 3*: EPA 1633 is capable of analyzing 40 compounds, our native standard solution only contains 30, so you will get errors that data for some compounds is not found. This is correct. If new ampoules are ordered for the other 10 compounds, everything needed for the calculation is there.
- *Note 4*: the FTS compounds only have 6 cal levels instead of 7.

In [7]:
from analogs_compounds import target_EIS_analogs

# Make a copy of rawdata to avoid modifying the original
rawdata_copy = rawdata.copy()
columns_to_rename = {}
for col in rawdata_copy.columns:
    new_col = col
    # Fix PFPHpA -> PFHpA
    if 'PFPHpA' in col:
        new_col = new_col.replace('PFPHpA', 'PFHpA')
    # Remove spaces from FTS compounds (e.g., "4:2 FTS" -> "4:2FTS")
    if ' FTS' in col:
        new_col = new_col.replace(' FTS', 'FTS')
    
    if new_col != col:
        columns_to_rename[col] = new_col

if columns_to_rename:
    rawdata_copy.rename(columns=columns_to_rename, inplace=True)

RF_target_analytes= calculate_average_RRs_targets(rawdata_copy, expected_concs_EIS, calibration_solutions, target_EIS_analogs)

Found 40 total target-IS pairs:
  NMeFOSE -> D7-NMeFOSE
  NEtFOSE -> D9-NEtFOSE
  PFBA -> 13C4-PFBA
  HFPO-DA -> 13C3-HFPO-DA
  PFPeA -> 13C5-PFPeA
  4:2FTS -> 13C2-4:2FTS
  6:2FTS -> 13C2-6:2FTS
  8:2FTS -> 13C2-8:2FTS
  NMeFOSAA -> D3-NMeFOSAA
  NEtFOSAA -> D5-NEtFOSAA
  PFHxA -> 13C5-PFHxA
  PFHpA -> 13C4-PFHpA
  PFOA -> 13C8-PFOA
  PFBS -> 13C3-PFBS
  PFHxS -> 13C3-PFHxS
  PFOS -> 13C8-PFOS
  PFOSA -> 13C8-PFOSA
  NMeFOSA -> D3-NMeFOSA
  NEtFOSA -> D5-NEtFOSA
  PFNA -> 13C9-PFNA
  PFDA -> 13C6-PFDA
  PFUnA -> 13C7-PFUnA
  PFDoA -> 13C2-PFDoA
  PFTeDA -> 13C2-PFTeDA
  PFTrDA -> 13C2-PFTeDA
  PFPeS -> 13C3-PFHxS
  PFHpS -> 13C8-PFOS
  PFNS -> 13C8-PFOS
  PFDS -> 13C8-PFOS
  PFDoS -> 13C8-PFOS
  ADONA -> 13C3-HFPO-DA
  PFMPA -> 13C5-PFPeA
  PFMBA -> 13C5-PFPeA
  NFDHA -> 13C5-PFHxA
  9Cl-PF3ONS -> 13C3-HFPO-DA
  11Cl-PF3OUdS -> 13C3-HFPO-DA
  PFEESA -> 13C5-PFHxA
  3:3FTCA -> 13C5-PFPeA
  5:3FTCA -> 13C5-PFHxA
  7:3FTCA -> 13C5-PFHxA


AVERAGE RESPONSE RATIOS FOR TARGET COMPOUNDS WITH

Target_Compound   EIS_Isotope  ...  N_Levels                 Levels_Used
0          NMeFOSE    D7-NMeFOSE  ...         7  L1, L2, L3, L4, L5, L6, L7
1          NEtFOSE    D9-NEtFOSE  ...         7  L1, L2, L3, L4, L5, L6, L7
2             PFBA     13C4-PFBA  ...         7  L1, L2, L3, L4, L5, L6, L7
3          HFPO-DA  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7
4            PFPeA    13C5-PFPeA  ...         7  L1, L2, L3, L4, L5, L6, L7
5           4:2FTS   13C2-4:2FTS  ...         6      L1, L2, L3, L4, L5, L6
6           6:2FTS   13C2-6:2FTS  ...         6      L1, L2, L3, L4, L5, L6
7           8:2FTS   13C2-8:2FTS  ...         6      L1, L2, L3, L4, L5, L6
8         NMeFOSAA   D3-NMeFOSAA  ...         7  L1, L2, L3, L4, L5, L6, L7
9         NEtFOSAA   D5-NEtFOSAA  ...         7  L1, L2, L3, L4, L5, L6, L7
10           PFHxA    13C5-PFHxA  ...         7  L1, L2, L3, L4, L5, L6, L7
11           PFHpA    13C4-PFHpA  ...         7  L1, L2, L3, L4, L5, L6, L7
12            PFOA     13C8-PFOA  ...         7  L1, L2, L3, L4, L5, L6, L7
13            PFBS     13C3-PFBS  ...         7  L1, L2, L3, L4, L5, L6, L7
14           PFHxS    13C3-PFHxS  ...         7  L1, L2, L3, L4, L5, L6, L7
15            PFOS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
16           PFOSA    13C8-PFOSA  ...         7  L1, L2, L3, L4, L5, L6, L7
17         NMeFOSA    D3-NMeFOSA  ...         7  L1, L2, L3, L4, L5, L6, L7
18         NEtFOSA    D5-NEtFOSA  ...         7  L1, L2, L3, L4, L5, L6, L7
19            PFNA     13C9-PFNA  ...         7  L1, L2, L3, L4, L5, L6, L7
20            PFDA     13C6-PFDA  ...         7  L1, L2, L3, L4, L5, L6, L7
21           PFUnA    13C7-PFUnA  ...         7  L1, L2, L3, L4, L5, L6, L7
22           PFDoA    13C2-PFDoA  ...         7  L1, L2, L3, L4, L5, L6, L7
23          PFTeDA   13C2-PFTeDA  ...         7  L1, L2, L3, L4, L5, L6, L7
24          PFTrDA   13C2-PFTeDA  ...         7  L1, L2, L3, L4, L5, L6, L7
25           PFPeS    13C3-PFHxS  ...         7  L1, L2, L3, L4, L5, L6, L7
26           PFHpS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
27            PFNS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
28            PFDS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
29           ADONA  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7
30      9Cl-PF3ONS  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7
31    11Cl-PF3OUdS  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7

[32 rows x 8 columns]

### 4.2 Calculating Concentrations target analytes
Here we compute the concentration of the target analytes (ng/L) based on the formula on page 44 of EPA method 1633. 

*Note 1* % solids is not added yet, needs to be done before the solids are added. 


In [31]:

conc_target_analytes = calculate_conc_targets(rawdata_copy,
                           RF_target_analytes,
                           EIS_stock, dilution_EIS_stock, spiked_amount_EIS, Ws,
                           DF_sample,  debug_compound=None )

Name    Type  ...  9Cl-PF3ONS_Conc_Calc  11Cl-PF3OUdS_Conc_Calc
0           blank + EIS  Sample  ...              0.000000                0.012141
1                QC std  Sample  ...            111.748518              113.542192
2                   KCS  Sample  ...             97.586505               94.680232
3                   DWZ  Sample  ...              0.001545                0.008544
4                   KWZ  Sample  ...              0.004931                0.009088
5                   DWP  Sample  ...            115.605196               95.840190
6                   KWP  Sample  ...             95.516111               78.988097
7                   KCS  Sample  ...             92.789094               90.604086
8                 QCstd  Sample  ...            109.461237              108.304356
9                 rinse  Sample  ...             31.024046               57.179464
10                rinse  Sample  ...            124.096183               57.179464
11                rinse  Sample  ...             41.365394               57.179464
12  blank_spike_eis-nis  Sample  ...              0.000000                0.004781
13                1_ccf  Sample  ...              0.006463                0.000000
14                 2_cf  Sample  ...              0.003284                0.012107
15                  3_f  Sample  ...              0.008029                0.004933
16               4_none  Sample  ...              0.002839                0.005233
17                rinse  Sample  ...                   inf                     NaN
25                RINSE  Sample  ...             31.024046               28.589732
26          blank_spike  Sample  ...              0.004970                0.004580
27             alex_kwz  Sample  ...              0.002280                0.008403
28             alex_dwz  Sample  ...              0.003181                0.005862
29          alex_kwpfas  Sample  ...            105.787820               92.530578
30          alex_dwpfas  Sample  ...            112.487890              105.742909

[24 rows x 34 columns]

## 5. Save to Excel

In [32]:

# Define the Excel file path
excel_file_path = "results/EPA_1633_results.xlsx"
# Save DataFrames to separate sheets
with pd.ExcelWriter(excel_file_path, engine='openpyxl') as writer:
    # 2.1 RSE EIS compounds
    df_rse_eis.to_excel(writer, sheet_name="2.1_RSE_EIS_Compounds", index=False)
    # 2.2 RSE target analytes
    df_rse_target.to_excel(writer, sheet_name="2.2_RSE_target", index=False)
    # 3.1 Concentration EIS compounds
    eis_RFS.to_excel(writer, sheet_name="3.1.1_RF_Eis", index=False)
    calculated_cons_EIS.to_excel(writer, sheet_name="3.1.2_concentration_EIS", index=False)

    #3.2 Recoveries EIS compounds
    calculated_recoveries_EIS.to_excel(writer, sheet_name="3.2_recoveries_EIS", index=False)

    #4.1 Target analytes response factors 
    RF_target_analytes.to_excel(writer, sheet_name="4.1 Target_analytes_RF", index=False)
    # 4.2 Target analytes concentrations
    conc_target_analytes.to_excel(writer, sheet_name="4.2_Target_Conc", index=False)

print(f"DataFrames saved to {excel_file_path}")


DataFrames saved to results/EPA_1633_results.xlsx
